File: backend/src/main/java/com/bookbook/domain/user/service/UserService.java
package com.bookbook.domain.user.service;

import com.bookbook.domain.review.repository.ReviewRepository;
import com.bookbook.domain.user.dto.UserProfileResponseDto;
import com.bookbook.domain.user.dto.UserResponseDto;
import com.bookbook.domain.user.dto.UserStatusResponseDto;
import com.bookbook.domain.user.entity.User;
import com.bookbook.domain.user.enums.Role;
import com.bookbook.domain.user.enums.UserStatus;
import com.bookbook.domain.user.repository.UserRepository;
import com.bookbook.global.exception.ServiceException;
import jakarta.annotation.PostConstruct;
import lombok.RequiredArgsConstructor;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class UserService {
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JdbcTemplate jdbcTemplate;
    private final ReviewRepository reviewRepository;

    @PostConstruct
    @Transactional
    public void initializeData(){
        createAdminUser();
        updateNotificationSchema();
    }

    private void createAdminUser(){
        String adminUsername = "admin";
        String adminPassword = "admin123";
        String adminEmail = "admin@book.com";

        if(!userRepository.existsByUsername(adminUsername)){
            User adminUser = User.builder()
                    .username(adminUsername)
                    .password(passwordEncoder.encode(adminPassword))
                    .email(adminEmail)
                    .nickname("관리자")
                    .address("서울시 강남구")
                    .rating(5.0f)
                    .role(Role.ADMIN)
                    .userStatus(UserStatus.ACTIVE)
                    .registrationCompleted(true)
                    .build();
            userRepository.save(adminUser);
            System.out.println("관리자 계정이 생성되었습니다: " + adminUsername);
        } else {
            System.out.println("관리자 계정이 이미 존재합니다: " + adminUsername);
        }
    }

    private void updateNotificationSchema() {
        try {
            // 기존 CHECK 제약 조건을 제거하고 새로운 것으로 교체
            jdbcTemplate.execute("ALTER TABLE notification DROP CONSTRAINT IF EXISTS notification_type_check");
            jdbcTemplate.execute("ALTER TABLE notification ADD CONSTRAINT notification_type_check CHECK (type IN ('RENT_REQUEST', 'RETURN_REMINDER', 'WISHLIST_AVAILABLE', 'POST_CREATED'))");
            System.out.println("✅ NotificationType 스키마 업데이트 완료: POST_CREATED 추가됨");
        } catch (Exception e) {
            System.err.println("❌ NotificationType 스키마 업데이트 실패: " + e.getMessage());
            // 개발 환경에서는 계속 진행
        }
    }

    public boolean checkNicknameAvailability(String nickname) {
        return !userRepository.existsByNickname(nickname);
    }

    @Transactional
    public void completeRegistration(Long userId, String nickname, String address) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ServiceException("404-USER-NOT-FOUND", "해당 ID의 사용자를 찾을 수 없습니다."));

        if (nickname == null || nickname.trim().isEmpty()) {
            throw new ServiceException("400-NICKNAME-EMPTY", "닉네임은 필수 입력 사항입니다.");
        }
        if (userRepository.existsByNickname(nickname.trim())) {
            throw new ServiceException("409-NICKNAME-DUPLICATE", "이미 사용 중인 닉네임입니다.");
        }
        if (address == null || address.trim().isEmpty()) {
            throw new ServiceException("400-ADDRESS-EMPTY", "주소는 필수 입력 사항입니다.");
        }

        user.setNickname(nickname.trim());
        user.setAddress(address.trim());
        user.setRegistrationCompleted(true);
        userRepository.save(user);
    }

    @Transactional
    public void updateUserInfo(Long userId, String nickname, String address) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ServiceException("404-USER-NOT-FOUND", "해당 ID의 사용자를 찾을 수 없습니다."));

        boolean hasChanges = false;

        if (nickname != null && !nickname.trim().isEmpty()) {
            String trimmedNickname = nickname.trim();
            if (!trimmedNickname.equals(user.getNickname())) {
                if (userRepository.existsByNickname(trimmedNickname)) {
                    throw new ServiceException("409-NICKNAME-DUPLICATE", "이미 사용 중인 닉네임입니다.");
                }
                user.setNickname(trimmedNickname);
                hasChanges = true;
            }
        }


        if (address != null && !address.trim().isEmpty()) {
            String trimmedAddress = address.trim();
            if (!trimmedAddress.equals(user.getAddress())) {
                user.setAddress(trimmedAddress);
                hasChanges = true;
            }
        }

        if (!hasChanges) {
            throw new ServiceException("400-NO-CHANGES", "변경할 닉네임 또는 주소를 제공하지 않았거나 변경사항이 없습니다.");
        }

        user.setRegistrationCompleted(true);
        userRepository.save(user);
    }

    public User findById(Long id) {
        return userRepository.findById(id).orElse(null);
    }

    @Transactional
    public void deactivateUser(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ServiceException("404-USER-NOT-FOUND", "해당 ID의 사용자를 찾을 수 없습니다."));

        if (user.getUserStatus() == UserStatus.INACTIVE) {
            throw new ServiceException("409-USER-ALREADY-INACTIVE", "이미 탈퇴 처리된 사용자입니다.");
        }

        user.changeUserStatus(UserStatus.INACTIVE);
        user.changeUsername(user.getUsername() + "_deleted_" + UUID.randomUUID().toString());
        if (user.getEmail() != null && !user.getEmail().isEmpty()) {
            user.setEmail(user.getEmail() + "_deleted_" + UUID.randomUUID().toString());
        }

        user.setNickname(null);
        user.setAddress(null);
        user.setRegistrationCompleted(false);

        userRepository.save(user);
    }

    public UserResponseDto getUserDetails(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ServiceException("404-USER-NOT-FOUND", "해당 ID의 사용자를 찾을 수 없습니다."));
        return new UserResponseDto(user);
    }

    public UserStatusResponseDto getUserStatus(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ServiceException("404-USER-NOT-FOUND", "해당 ID의 사용자를 찾을 수 없습니다."));

        return UserStatusResponseDto.builder()
                .id(user.getId())
                .userStatus(user.getUserStatus())
                .build();
    }

    @Transactional
    public User findOrCreateUser(String username, String email, String socialNickname) {
        return userRepository.findByUsername(username)
                .map(user -> {
                    if (user.getEmail() == null && email != null) {
                        user.setEmail(email);
                    }
                    return user;
                })
                .orElseGet(() -> {
                    User newUser = User.builder()
                            .username(username)
                            .email(email)
                            .password(passwordEncoder.encode(UUID.randomUUID().toString()))
                            .nickname(null)
                            .address(null)
                            .rating(0.0f)
                            .role(Role.USER)
                            .userStatus(UserStatus.ACTIVE)
                            .registrationCompleted(false)
                            .build();
                    userRepository.save(newUser);
                    return newUser;
                });
    }

    public User getByIdOrThrow(Long id) {
        return userRepository.findById(id)
                .orElseThrow(() -> new ServiceException("404-USER-NOT-FOUND", "해당 사용자를 찾을 수 없습니다."));
    }

    public UserProfileResponseDto getUserProfileDetails(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ServiceException("404-USER-NOT-FOUND", "사용자를 찾을 수 없습니다."));

        if (user.getRole() == Role.ADMIN) {
            return UserProfileResponseDto.from(user, 5.0, 0);
        }
        
        Optional<Double> averageRating = reviewRepository.findAverageRatingByRevieweeId(userId);
        long mannerScoreCount = reviewRepository.countByRevieweeId(userId);

        double mannerScore = averageRating.orElse(0.0);

        return UserProfileResponseDto.from(user, mannerScore, (int) mannerScoreCount);
    }

}

File: backend/src/main/java/com/bookbook/domain/user/service/UserService.kt
package com.bookbook.domain.user.service

import com.bookbook.domain.review.repository.ReviewRepository
import com.bookbook.domain.user.dto.response.UserProfileResponseDto
import com.bookbook.domain.user.dto.response.UserResponseDto
import com.bookbook.domain.user.dto.response.UserStatusResponseDto
import com.bookbook.domain.user.entity.User
import com.bookbook.domain.user.enums.Role
import com.bookbook.domain.user.enums.UserStatus
import com.bookbook.domain.user.repository.UserRepository
import com.bookbook.global.exception.ServiceException
import jakarta.annotation.PostConstruct
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional(readOnly = true)
class UserService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jdbcTemplate: JdbcTemplate,
    private val reviewRepository: ReviewRepository
) {

    @PostConstruct
    @Transactional
    fun initializeData() {
        createAdminUser()
        updateNotificationSchema()
    }

    private fun createAdminUser() {
        val adminUsername = "admin"
        val adminPassword = "admin123"
        val adminEmail = "admin@book.com"

        if (!userRepository.existsByUsername(adminUsername)) {
            val adminUser = User(
                username = adminUsername,
                password = passwordEncoder.encode(adminPassword),
                email = adminEmail,
                nickname = "관리자",
                address = "서울시 강남구",
                role = Role.ADMIN,
                registrationCompleted = true
            )
            userRepository.save(adminUser)
            println("관리자 계정이 생성되었습니다: $adminUsername")
        } else {
            println("관리자 계정이 이미 존재합니다: $adminUsername")
        }
    }

    private fun updateNotificationSchema() {
        try {
            jdbcTemplate.execute("ALTER TABLE notification DROP CONSTRAINT IF EXISTS notification_type_check")
            jdbcTemplate.execute("ALTER TABLE notification ADD CONSTRAINT notification_type_check CHECK (type IN ('RENT_REQUEST', 'RENT_APPROVED', 'RENT_REJECTED', 'WISHLIST_AVAILABLE', 'RETURN_REMINDER', 'POST_CREATED'))")
            println("✅ NotificationType 스키마 업데이트 완료: 모든 enum 값 포함됨")
        } catch (e: Exception) {
            System.err.println("❌ NotificationType 스키마 업데이트 실패: ${e.message}")
        }
    }

    fun checkNicknameAvailability(nickname: String): Boolean {
        return !userRepository.existsByNickname(nickname)
    }

    @Transactional
    fun completeRegistration(userId: Long, nickname: String?, address: String?) {
        val user = getByIdOrThrow(userId)

        val trimmedNickname = nickname?.trim()
        val trimmedAddress = address?.trim()

        if (trimmedNickname.isNullOrEmpty()) {
            throw ServiceException("400-NICKNAME-EMPTY", "닉네임은 필수 입력 사항입니다.")
        }
        if (userRepository.existsByNickname(trimmedNickname)) {
            throw ServiceException("409-NICKNAME-DUPLICATE", "이미 사용 중인 닉네임입니다.")
        }
        if (trimmedAddress.isNullOrEmpty()) {
            throw ServiceException("400-ADDRESS-EMPTY", "주소는 필수 입력 사항입니다.")
        }

        user.nickname = trimmedNickname
        user.address = trimmedAddress
        user.registrationCompleted = true
        userRepository.save(user)
    }

    @Transactional
    fun updateUserInfo(userId: Long, nickname: String?, address: String?) {
        val user = getByIdOrThrow(userId)

        val trimmedNickname = nickname?.trim()
        val trimmedAddress = address?.trim()

        var hasChanges = false

        if (!trimmedNickname.isNullOrEmpty() && trimmedNickname != user.nickname) {
            if (userRepository.existsByNickname(trimmedNickname)) {
                throw ServiceException("409-NICKNAME-DUPLICATE", "이미 사용 중인 닉네임입니다.")
            }
            user.nickname = trimmedNickname
            hasChanges = true
        }

        if (!trimmedAddress.isNullOrEmpty() && trimmedAddress != user.address) {
            user.address = trimmedAddress
            hasChanges = true
        }

        if (!hasChanges) {
            throw ServiceException("400-NO-CHANGES", "변경할 닉네임 또는 주소를 제공하지 않았거나 변경사항이 없습니다.")
        }

        user.registrationCompleted = true
        userRepository.save(user)
    }

    fun findById(id: Long): User? {
        return userRepository.findById(id).orElse(null)
    }

    @Transactional
    fun deactivateUser(userId: Long) {
        val user = getByIdOrThrow(userId)

        if (user.userStatus == UserStatus.INACTIVE) {
            throw ServiceException("409-USER-ALREADY-INACTIVE", "이미 탈퇴 처리된 사용자입니다.")
        }

        user.apply {
            userStatus = UserStatus.INACTIVE
            username = "${username}_deleted_${UUID.randomUUID()}"
            email = email?.let { "${it}_deleted_${UUID.randomUUID()}" }
            nickname = null
            address = null
            registrationCompleted = false
        }

        userRepository.save(user)
    }

    fun getUserDetails(userId: Long): UserResponseDto {
        val user = getByIdOrThrow(userId)
        return UserResponseDto(user)
    }

    fun getUserStatus(userId: Long): UserStatusResponseDto {
        val user = getByIdOrThrow(userId)
        return UserStatusResponseDto(user.id, user.userStatus)
    }

    @Transactional
    fun findOrCreateUser(username: String, email: String?, socialNickname: String?): User {
        return userRepository.findByUsername(username)
            .orElseGet {
                val newUser = User(
                    username = username,
                    email = email,
                    password = passwordEncoder.encode(UUID.randomUUID().toString()),
                )
                userRepository.save(newUser)
                newUser
            }
    }

    fun getByIdOrThrow(id: Long): User {
        return userRepository.findById(id)
            .orElseThrow { ServiceException("404-USER-NOT-FOUND", "해당 사용자를 찾을 수 없습니다.") }
    }

    fun getUserProfileDetails(userId: Long): UserProfileResponseDto {
        val user = getByIdOrThrow(userId)

        return if (user.isAdmin) {
            UserProfileResponseDto(user, 5.0, 0)
        } else {
            val averageRating = reviewRepository.findAverageRatingByRevieweeId(userId).orElse(0.0)
            val mannerScoreCount = reviewRepository.countByRevieweeId(userId)
            UserProfileResponseDto(user, averageRating, mannerScoreCount.toInt())
        }
    }
}